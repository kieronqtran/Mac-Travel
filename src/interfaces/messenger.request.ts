export interface MessengerRequest {
  object: 'page';
  entry: Entry[];
}

export interface Entry {
  id: string;
  time: number;
  messaging: Messaging[];
}

export interface Messaging {
  sender: Sender;
  recipient: Recipient;
  timestamp: number;
  message: Message;
}

export interface Sender {
  id: string;
}

export interface Recipient {
  id: string;
}

export interface Payload {
  url: string;
  sticker_id: number;
}

export interface Attachment {
  type: string;
  payload: Payload;
}

export interface Message {
  mid: string;
  seq: number;
  sticker_id: number;
  attachments: Attachment[];
}
