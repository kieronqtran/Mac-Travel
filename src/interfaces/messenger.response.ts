interface MessengerResponse {
  recipient: { id: string };
}

interface QuickReplyResponse extends MessengerResponse {
  message: {
    text: string;
    quick_replies: [
      {
        content_type: 'text';
        title: string;
        payload: string;
      }
    ];
  };
}
