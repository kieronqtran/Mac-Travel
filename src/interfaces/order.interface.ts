import { Document } from 'mongoose';
import { User, Product } from './index';

export interface OrderDetail {
  quantity: number;
  total_price: number;
  product: Product;
}

export interface ShippingAddress {
  customer_name: string;
  city: string;
  postal_code: string;
  state: string;
  street: string;
  country: string;
}

export interface Order extends Document {
  payment_type: string;
  currency: string;
  subtotal: number;
  shipping_cost: number;
  total_cost: number;
  total_tax: number;
  checkouted: boolean;
  checkoutTime: string;
  updatedTime: number;
  createdTime: number;
  order_user: User;
  order_details: OrderDetail[];
  shipping_address: ShippingAddress;
}
