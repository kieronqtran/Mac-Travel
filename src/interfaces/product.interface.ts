import { Document } from 'mongoose';

export interface Product extends Document {
  name: string;
  type: string;
  item_url: string;
  image_url: string;
  unit_price: number;
  currency: string;
  description: string;
  payload_name: string;
}
