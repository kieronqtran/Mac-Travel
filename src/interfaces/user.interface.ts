import { Document } from 'mongoose';

export interface User extends Document {
  facebook_id: string;
  first_name: string;
  last_name: string;
  timezone: number;
  gender: string;
}
