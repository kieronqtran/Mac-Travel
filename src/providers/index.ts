export * from './user.provider';
export * from './product.provider';
export * from './order.provider';
export * from './database.provider';
