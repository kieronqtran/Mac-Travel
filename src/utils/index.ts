export * from './config';
export * from './intent.enum';
