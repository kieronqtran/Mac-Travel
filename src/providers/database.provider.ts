import * as mongoose from 'mongoose';
import { Mockgoose } from 'mockgoose-fix';
import { environments } from '../utils';
export let mockgoose: Mockgoose;
if (process.env.NODE_ENV === 'test') {
  mockgoose = new Mockgoose(mongoose);
}

export const databaseProvider = {
  provide: 'DbConnection',
  useFactory: async () => {
    (mongoose as any).Promise = global.Promise;

    if (process.env.NODE_ENV === 'test') {
      mockgoose.helper.setDbVersion('3.4.14');
      await mockgoose.prepareStorage();
      return await mongoose.connect('mongodb://localhost:27017/mactravel');
    } else {
      return await mongoose.connect(environments.mongoUrl);
    }
  },
};
