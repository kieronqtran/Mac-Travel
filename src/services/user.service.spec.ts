import { Test } from '@nestjs/testing';
import { Connection, Model } from 'mongoose';
import { UserService } from './user.service';
import { User } from '../interfaces';
import { userProvider, databaseProvider, mockgoose } from '../providers';
import { environments } from '../utils';

const fetch: any = require('node-fetch');

describe('UserService', () => {
  let userService: UserService;
  let user: Model<User>;

  const expectedUser = {
    facebook_id: '2042621042422137',
    first_name: 'Quang',
    last_name: 'Tran',
    gender: 'male',
    timezone: 7,
  };

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      components: [UserService, databaseProvider, userProvider],
    }).compile();

    userService = module.get(UserService);
    user = module.get('UserModelToken');
    fetch.resetMocks();
  });

  afterEach(async () => {
    await mockgoose.helper.reset();
    jest.resetAllMocks();
  });

  describe('getUserByFacebookId', () => {
    test('should fetch user if not exist', async () => {
      fetch.mockImplementation(async () => ({
        json: async () => Promise.resolve({
          id: expectedUser.facebook_id,
          first_name: expectedUser.first_name,
          last_name: expectedUser.last_name,
          gender: expectedUser.gender,
          timezone: expectedUser.timezone,
        })));
      const beforeLength = await user.count({}).exec();
      const actualUser = await userService.getUserByFacebookId(expectedUser.facebook_id);
      expect(fetch.mock.calls).toMatchSnapshot();
      expect(actualUser.facebook_id).toBe(expectedUser.facebook_id);
      expect(actualUser.first_name).toBe(expectedUser.first_name);
      expect(actualUser.last_name).toBe(expectedUser.last_name);
      expect(actualUser.gender).toBe(expectedUser.gender);
      expect(actualUser.timezone).toBe(expectedUser.timezone);
      expect(await user.count({}).exec()).toEqual(beforeLength + 1);
    });

    test('should get a facebook user', async () => {
      const model = new user(expectedUser);
      await model.save();
      const beforeLength = await user.count({}).exec();
      const actualUser = await userService.getUserByFacebookId(expectedUser.facebook_id);
      expect(actualUser.facebook_id).toBe(expectedUser.facebook_id);
      expect(actualUser.first_name).toBe(expectedUser.first_name);
      expect(actualUser.last_name).toBe(expectedUser.last_name);
      expect(actualUser.gender).toBe(expectedUser.gender);
      expect(actualUser.timezone).toBe(expectedUser.timezone);
      expect(await user.count({}).exec()).toEqual(beforeLength)
    });
  });
});
