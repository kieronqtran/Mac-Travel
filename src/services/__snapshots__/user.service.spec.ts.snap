// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService getUserByFacebookId should fetch user if not exist 1`] = `
Array [
  Array [
    "https://graph.facebook.com/v2.12/me/2042621042422137?access_token=this_is_a_fake_token&fields=%27first_name%2Clast_name%2Ctimezone%2Cgender%27",
    Object {
      "headers": Object {
        "Accept": "application/json",
        "Content-Type": "application/json",
      },
      "method": "GET",
    },
  ],
]
`;
